### PR TITLE
Integrate zebra ophyd device

### PR DESCRIPTION
**Pull request by JunAishima**
refs #13 

Apologies, didn't mean to push to upstream, but it's become a habit recently!

### CODE REVIEW

**Comment by mrakitin** [comment]
I thought we were about to push these changes to https://github.com/NSLS-II/lsdc/tree/scans-ophyd-bluesky after merging, not `master` (as documented in https://github.com/NSLS-II/lsdc/projects). I changed the base branch.

**Review by mrakitin — CHANGES_REQUESTED** [review]
Some suggestions to make it compatible with the way we coded it in #28.

**Comment by JunAishima** [comment]
sorry, will go back and push correctly

**Comment by mrakitin** [comment]
@JunAishima, no worries. I have changed the target branch already:
image

File: zebra.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    encoder = Cpt(EpicsSignal, 'ENC')
+    encX = Cpt(EpicsSignal, 'ENC1')
+    encY = Cpt(EpicsSignal, 'ENC2')
+    encZ = Cpt(EpicsSignal, 'ENC3')
+    enzO = Cpt(EpicsSignal, 'ENC4')
+    encoder = Cpt(EpicsSignal, 'ENC')
+    direction = Cpt(EpicsSignal, 'DIR')
+    gate = Cpt(ZebraPCGate, 'GATE_')
+    pulse = Cpt(ZebraPCPulse, 'PULSE_')
+
```
[the remark this answers is not in the document]
Comment (JunAishima) [reply zebra.py:34]: have added blank lines at all appropriate places. should be clean now

File: zebra.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+from ophyd import (Component as Cpt, EpicsSignal, EpicsSignalRO, Device)
+
+class ZebraPCGate(Device):
+    sel = Cpt(EpicsSignal, 'SEL')
+    start = Cpt(EpicsSignal, 'START')
+    width = Cpt(EpicsSignal, 'WID')
+    step = Cpt(EpicsSignal, 'STEP')
```
[the remark this answers is not in the document]
Comment (JunAishima) [reply zebra.py:8]: have followed the request and used ZebraPCBase

File: zebra.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    enzO = Cpt(EpicsSignal, 'ENC4')
+    encoder = Cpt(EpicsSignal, 'ENC')
+    direction = Cpt(EpicsSignal, 'DIR')
+    gate = Cpt(ZebraPCGate, 'GATE_')
+    pulse = Cpt(ZebraPCPulse, 'PULSE_')
+
+class ZebraAnd(Device):
+    inp1 = Cpt(EpicsSignal, 'INP1:STA')
+    inp2 = Cpt(EpicsSignal, 'INP2:STA')
+    out1 = Cpt(EpicsSignal, 'OUT1_TTL') #out type?
```
[the remark this answers is not in the document]
Comment (JunAishima) [reply zebra.py]: removed comment

File: zebra.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    start = Cpt(EpicsSignal, 'START')
+    width = Cpt(EpicsSignal, 'WID')
+    step = Cpt(EpicsSignal, 'STEP')
+    max = Cpt(EpicsSignal, 'MAX')
+    delay = Cpt(EpicsSignal, 'DLY')
+    status = Cpt(EpicsSignalRO, 'INP.STA')
+
+class ZebraPositionCompare(Device):
+    armStatus = Cpt(EpicsSignalRO, 'ARM_INP.STA')
+    downloadCount = Cpt(EpicsSignalRO, 'NUM_DOWN')
```
[the remark this answers is not in the document]
Comment (JunAishima) [reply zebra.py]: done

File: zebra.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    inp2 = Cpt(EpicsSignal, 'INP2:STA')
+    out1 = Cpt(EpicsSignal, 'OUT1_TTL') #out type?
+
+class Zebra(Device):
+    downloading = Cpt(EpicsSignal, 'ARRAY_ARQ')
+    reset = Cpt(EpicsSignal, 'SYS_RESET.PROC')
+    m1SetPos = Cpt(EpicsSignal, 'M1:SETPOS.PROC')
+    m2SetPos = Cpt(EpicsSignal, 'M2:SETPOS.PROC')
+    m3SetPos = Cpt(EpicsSignal, 'M3:SETPOS.PROC')
+    m4SetPos = Cpt(EpicsSignal, 'M4:SETPOS.PROC')
```
[the remark this answers is not in the document]
Comment (JunAishima) [reply zebra.py]: done

**Comment by JunAishima** [comment]
I believe I have followed all of Maksim's change requests.

File: zebra.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+class ZebraPositionCompare(Device):
+    arm_status = Cpt(EpicsSignalRO, 'ARM_INP.STA')
+    download_count = Cpt(EpicsSignalRO, 'NUM_DOWN')
+    disarm = Cpt(EpicsSignal, 'DISARM')
+    encoder = Cpt(EpicsSignal, 'ENC')
+    enc_x = Cpt(EpicsSignal, 'ENC1')
+    enc_y = Cpt(EpicsSignal, 'ENC2')
+    enc_z = Cpt(EpicsSignal, 'ENC3')
+    enz_o = Cpt(EpicsSignal, 'ENC4')
```
Comment (mrakitin) [inline zebra.py:29]: Is it supposed to be `enz_o`? What does it stand for? It seems to be breaking the pattern established for the x/y/x PVs above.

**Review by mrakitin — COMMENTED** [review]
Thanks Jun, the code looks good to me now. I spotted a potential typo (see below), could you please confirm the name is expected?

File: zebra.py (hunk 6 above)
Comment (JunAishima) [reply zebra.py:29]: this does actually make sense - I can rename this to omega (I guess I just shortened this too much to "o" instead of "omega")

Comment (mrakitin) [reply zebra.py:29]: My concern was mostly about the `z` letter in `enz_o`. I think it should be `enc_o`.

Comment (JunAishima) [reply zebra.py:29]: oh! didn't even see that. thanks for catching it.

I still think that maybe o is a bit too terse... will submit a pull request with both changes

Comment (mrakitin) [inline zebra.py:29]:
```suggestion
    enc_omega = Cpt(EpicsSignal, 'ENC4')
```

File: zebra.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrakitin) [reply zebra.py:8]: Thanks!

**Review by mrakitin — APPROVED** [review]
Looks good to me now. I think it's ready to go in and used.

**Comment by JunAishima** [comment]
Hey @tacaswell , @MikeHart85 , @AbbyGi , could you review this change? I would like to merge it and start populating the scans-ophyd-bluesky branch. Thanks for your help!

File: zebra.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+class ZebraPositionCompare(Device):
+    arm_status = Cpt(EpicsSignalRO, 'ARM_INP.STA')
+    download_count = Cpt(EpicsSignalRO, 'NUM_DOWN')
+    disarm = Cpt(EpicsSignal, 'DISARM')
+    encoder = Cpt(EpicsSignal, 'ENC')
+    enc_x = Cpt(EpicsSignal, 'ENC1')
+    enc_y = Cpt(EpicsSignal, 'ENC2')
+    enc_z = Cpt(EpicsSignal, 'ENC3')
+    enc_omega = Cpt(EpicsSignal, 'ENC4')
+    encoder = Cpt(EpicsSignal, 'ENC')
```
Comment (AbbyGi) [inline zebra.py:30]: It looks like `encoder` was defined the same way twice in this class (lines 25 and 30).

**Review by AbbyGi — COMMENTED** [review]
Looks good. Just one thing I noticed that may need to be changed.

**Review by MikeHart85 — APPROVED** [review]
Looks good to me. Nice catch @AbbyGi.

File: zebra.py (hunk 7 above)
Comment (JunAishima) [reply zebra.py]: thanks, again! turns out I had not pushed the comment through when we talked at the meeting.

**Pull request merged**